src/global/global_init.cc

```diff
@@ -214,21 +214,32 @@ global_init(const std::map<std::string,std::string> *defaults,
     gid_t gid = 0;
     std::string uid_string;
     std::string gid_string;
+    std::string home_directory;
     if (g_conf()->setuser.length()) {
+      char buf[4096];
+      struct passwd pa;
+      struct passwd *p = 0;
+
       uid = atoi(g_conf()->setuser.c_str());
-      if (!uid) {
-	char buf[4096];
-	struct passwd pa;
-	struct passwd *p = 0;
+      if (uid) {
+        getpwuid_r(uid, &pa, buf, sizeof(buf), &p);
+      } else {
 	getpwnam_r(g_conf()->setuser.c_str(), &pa, buf, sizeof(buf), &p);
-	if (!p) {
-	  cerr << "unable to look up user '" << g_conf()->setuser << "'"
-	       << std::endl;
-	  exit(1);
-	}
+      }
+
+      if (!uid && !p) {
+	cerr << "unable to look up user '" << g_conf()->setuser << "'"
+	     << std::endl;
+	exit(1);
+      }
+
+      if (p) {
 	uid = p->pw_uid;
 	gid = p->pw_gid;
 	uid_string = g_conf()->setuser;
```

Comment:
Nit: I suppose now, if 'setuser' option is specified as numeric (uid) value, `uid_string` will be set to this value, while it is expected to contain the user name. The previous behavior in this case seemed to have `uid_string` empty. I am not sure what is better (as it seems to be used only for diagnostic), but I suppose we could improve this by using `p->pw_name` here?

Reply:
I tweaked it to resemble the old behaviour as closely as possible.